generate-offline-map-using-workmanager/src/main/java/com/esri/arcgismaps/sample/generateofflinemapusingworkmanager/OfflineJobWorker.kt

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2023 Esri
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.esri.arcgismaps.sample.generateofflinemapusingworkmanager
+
+import android.content.Context
+import android.util.Log
+import androidx.work.CoroutineWorker
+import androidx.work.WorkerParameters
+import androidx.work.ForegroundInfo
+import androidx.work.workDataOf
+import com.arcgismaps.tasks.JobStatus
+import com.arcgismaps.tasks.offlinemaptask.GenerateOfflineMapJob
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.CancellationException
+import java.io.File
+
+/**
+ * Class that runs a GenerateOfflineMapJob as a CoroutineWorker using WorkManager.
+ */
+class OfflineJobWorker(private val context: Context, params: WorkerParameters) :
+    CoroutineWorker(context, params) {
+
+    private val TAG = OfflineJobWorker::class.java.simpleName
+
+    // notificationId passed by the activity
+    private val notificationId by lazy {
+        inputData.getInt(notificationIdParameter, 1)
+    }
+
+    // WorkerNotification instance
+    private val workerNotification by lazy {
+        WorkerNotification(context, notificationId)
+    }
+
+    // must override for api versions < 31 for backwards compatibility
+    // with foreground services
+    override suspend fun getForegroundInfo(): ForegroundInfo {
+        return createForegroundInfo(0)
+    }
+
+    /**
+     * Creates and returns a new ForegroundInfo with a progress notification and the given
+     * [progress] value.
+     */
+    private fun createForegroundInfo(progress: Int): ForegroundInfo {
+        // create a ForegroundInfo using the notificationId and a new progress notification
+        return ForegroundInfo(
+            notificationId,
+            workerNotification.createProgressNotification(progress)
+        )
+    }
+
+    override suspend fun doWork(): Result {
+        // CoroutineScope to run the GenerateOfflineMapJob's state flows on
+        val coroutineScope = CoroutineScope(Dispatchers.Default)
```

Comment:
I think you are better off using [coroutineScope](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/coroutine-scope.html) function here